### PR TITLE
[ConstraintSystem] Compute bindings incrementally

### PR DESCRIPTION
**Pull request by xedin**
Instead of re-gathering constraints and computing a full new set bindings on every step of the solver,
let's keep track of _potential_ bindings as constraint system is modified (new constraints are added/removed,
equivalence classes changed and type variables are bound/unbound).
 
`PotentialBindings` lost most of its responsibilities and are no longer comparable, their main
purpose now is binding  and metadata tracking (introduction/retraction). New `BindingSet` type 
is something that represents a set of bindings at the current step of the solver.

`ConstraintGraph` has been adjusted to  split "fixed bindings" into:
  - Referenced by: The set of type variables that are referenced by a fixed type associated with
     current type variable;
  - References: The set of type variables that reference current type variable in their fixed types.

That makes it easier to compute a set of type variables that are affected by any given change in
constraint graph.

### CODE REVIEW

**Comment by xedin** [comment]
@swift-ci please test source compatibility

**Comment by xedin** [comment]
@swift-ci please smoke test compiler performance

**Comment by xedin** [comment]
4 UPasses in the suite.

**Comment by xedin** [comment]
My suspicion is this is due to changes I made to `isViable` that are more relax now when it comes to filtering types with type variables...

**Comment by xedin** [comment]
@swift-ci please smoke test compiler performance

**Comment by xedin** [comment]
Ok, so revering `isViable` changes remove 20 seconds vs. previous build although it looks like there is still some more conformance checking going on. I'm pretty sure it's related to literal handling.

**Comment by xedin** [comment]
@swift-ci please smoke test compiler performance

**Comment by xedin** [comment]
@swift-ci please smoke test compiler performance

**Comment by xedin** [comment]
Hm, looks like there is something else going on here. This 55% difference is too big to be explained by just literal requirement handling...

**Comment by xedin** [comment]
@swift-ci please smoke test compiler performance

**Comment by xedin** [comment]
@swift-ci please smoke test compiler performance

**Comment by xedin** [comment]
@swift-ci please smoke test compiler performance

**Comment by xedin** [comment]
@swift-ci please test compiler performance

**Comment by xedin** [comment]
@swift-ci please smoke test compiler performance

**Comment by xedin** [comment]
@swift-ci please test compiler performance

**Comment by xedin** [comment]
Finally!

**Comment by xedin** [comment]
Great, full suite is green!

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please test source compatibility debug

**Comment by xedin** [comment]
This version is not ideal from performance perspective but I think I can fix extraneous conformance checks in the solver itself.

**Comment by xedin** [comment]
Since the underlying problem is that dependent member types are converted back to original DependentMemberType even though base doesn't conform to expected protocol which means that constraint graph gets notified about the binding as-if it was correct.

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please test source compatibility debug

**Comment by xedin** [comment]
@swift-ci please test Linux platform

**Comment by xedin** [comment]
@swift-ci please test macOS platform

File: lib/Sema/CSBindings.cpp [hunk 1]
```diff
@@ -481,8 +481,173 @@ void PotentialBindings::finalize(
...
+      if (!isCovered)
+        continue;
+
+      binding = binding.withType(adjustedTy);
+      info.setCoveredBy(binding.getSource());
+    }
+  }
+
+  Bindings.insert(std::move(binding));
+>>>>>>> [CSBindings] Separate inference from final product
```
Comment (hborla) [inline lib/Sema/CSBindings.cpp]: Looks like a git conflict marker left in this commit?

**Review by hborla — APPROVED** [review]
This looks good to me, and I really like the way the new code is structured and documented!

File: lib/Sema/CSBindings.cpp (hunk 1 above)
Comment (xedin) [reply lib/Sema/CSBindings.cpp]: Ah, it might be old code which I fixed :)

**Pull request merged**